core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,332 @@
+            return Err(AdviceSetError::InvalidPathLength(self.total_depth, depth));
+        }
+
+        let pos = 2u64.pow(self.total_depth) + index;
+        let index = pos / 2;
+
+        let mut path = self.paths[&index].clone();
+
+        if is_even(pos) {
+            path.remove(0);
+            Ok(path)
+        } else {
+            path.remove(1);
+            Ok(path)
+        }
+    }
+
+    /// Replaces the leaf at the specified index with the provided value.
+    ///
+    /// # Errors
+    /// Returns an error if the node with specified index is not in the map.
+    pub fn update_leaf(&mut self, index: u64, value: Word) -> Result<(), AdviceSetError> {
+        let depth = self.total_depth;
+        if index >= 2u64.pow(depth) {
+            return Err(AdviceSetError::InvalidIndex(depth, index));
+        }
+
+        // Fill old_hashes vector -----------------------------------------------------------------
+        let pos = 2u64.pow(depth) + index;
+        let path = self.paths.get(&(pos / 2)).unwrap().clone();
+
+        let old_hashes = self.compute_hashes(&path, depth, index, false);
+
+        // Fill new_hashes vector -----------------------------------------------------------------
+        let pos = 2u64.pow(depth) + index;
+
+        if is_even(pos) {
+            self.paths.get_mut(&(pos / 2)).unwrap()[0] = value;
+        } else {
+            self.paths.get_mut(&(pos / 2)).unwrap()[1] = value;
+        }
+
+        let path = self.paths.get(&(pos / 2)).unwrap().clone();
+
+        let new_hashes = self.compute_hashes(&path, depth, index, true);
+
+        // update paths ---------------------------------------------------------------------------
+        for (_, path) in self.paths.iter_mut() {
+            for i in (0..old_hashes.len()).rev() {
+                if path[i + 2] == old_hashes[i] {
+                    path[i + 2] = new_hashes[i];
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn compute_hashes(
+        &mut self,
+        path: &[Word],
+        depth: u32,
+        index: u64,
+        update_root: bool,
+    ) -> Vec<Word> {
+        let mut pos = 2u64.pow(depth) + index;
+        let mut path_iter = path.iter();
+
+        let mut computed_hashes = Vec::<Word>::new();
+
+        let mut path_hash;
+        let mut comp_hash;
+
+        if is_even(pos) {
+            comp_hash = *path_iter.next().unwrap();
+            path_hash = *path_iter.next().unwrap();
+        } else {
+            path_hash = *path_iter.next().unwrap();
+            comp_hash = *path_iter.next().unwrap();
+        }
+
+        for _ in 0..depth - 2 {
+            comp_hash = calculate_parent_hash(comp_hash, pos, path_hash);
+            pos /= 2;
+            computed_hashes.push(comp_hash);
+            path_hash = *path_iter.next().unwrap();
+        }
+
+        if update_root {
+            self.root = calculate_parent_hash(comp_hash, pos, path_hash);
```

Comment:
I wonder if it would be cleaner to mutate the root not here, but in the place where this function is called from.